implementations/rust/ockam/ockam_ffi/src/vault.rs

```diff
@@ -2,22 +2,67 @@ use crate::vault_types::{FfiSecretAttributes, SecretKeyHandle};
 use crate::{check_buffer, FfiError, FfiOckamError};
 use crate::{FfiVaultFatPointer, FfiVaultType};
 use core::{future::Future, result::Result as StdResult, slice};
+use futures::future::join_all;
 use lazy_static::lazy_static;
+use ockam_core::compat::collections::BTreeMap;
 use ockam_core::compat::sync::Arc;
 use ockam_core::vault::{
-    AsymmetricVault, Hasher, PublicKey, Secret, SecretAttributes, SecretVault, SymmetricVault,
+    AsymmetricVault, Hasher, KeyId, PublicKey, SecretAttributes, SecretVault, SymmetricVault,
 };
 use ockam_core::{Error, Result};
 use ockam_vault::Vault;
 use tokio::{runtime::Runtime, sync::RwLock, task};
 
-/// FFI Vault trait. See documentation for individual sub-traits for details.
-pub trait FfiVault: SecretVault + Hasher + SymmetricVault + AsymmetricVault + Send {}
+#[derive(Default)]
+struct SecretsMapping {
+    mapping: BTreeMap<u64, KeyId>,
+    last_index: u64,
+}
+
+impl SecretsMapping {
+    fn insert(&mut self, key_id: KeyId) -> u64 {
+        self.last_index += 1;
+
+        self.mapping.insert(self.last_index, key_id);
+
+        self.last_index
+    }
+
+    fn get(&self, index: u64) -> Result<KeyId> {
+        Ok(self
+            .mapping
+            .get(&index)
+            .cloned()
+            .ok_or(FfiError::EntryNotFound)?)
+    }
+
+    fn take(&mut self, index: u64) -> Result<KeyId> {
+        Ok(self.mapping.remove(&index).ok_or(FfiError::EntryNotFound)?)
+    }
+}
+
+#[derive(Clone, Default)]
+struct VaultEntry {
+    vault: Vault,
+    secrets_mapping: Arc<RwLock<SecretsMapping>>,
+}
+
+impl VaultEntry {
+    async fn insert(&self, key_id: KeyId) -> u64 {
+        self.secrets_mapping.write().await.insert(key_id)
```

Comment:
This doesn't really matter, since I'm going to delete the whole ffi crate very soon (and replace it with tcp side-car), and I wanted to spend minimum time on it just to keep it working. But what are the benefits of having sync RwLock here?

Reply:
> This doesn't really matter, since I'm going to delete the whole ffi crate very soon [...]

:ok: 

> But what are the benefits of having sync RwLock here?

Usually async locks are necessary only if the lock guard needs to be used across await points, or if there is very high contention and no viable sharding strategy exists. OTOH regular locks are much faster.